### ISSUE TITLE
JsonSerializer doesn't support valid TimeSpan formats

### ISSUE THREAD

**Issue by eerhardt** [dotnet/runtime#100538]
### Description

When trying to deserialize TimeSpan strings from JSON into an object using `JsonSerializer.Deserialize`, not all valid TimeSpan formats are supported.

This is inconsistent with our other APIs that deserialize TimeSpan values from strings (for example `Microsoft.Extensions.Configuration.Binder`). It can cause problems when people are reading TimeSpan information from other sources and can't control the format being written.

### Reproduction Steps

```C#
using Newtonsoft.Json;
using System.Buffers.Text;
using System.Text;

string[] timeSpanStrings = ["00:01", "10"];

foreach (string timeSpanString in timeSpanStrings)
{
    TimeSpan timeSpan = TimeSpan.Parse(timeSpanString);
    Console.WriteLine(timeSpan);

    var timeSpanStringUtf8 = Encoding.UTF8.GetBytes(timeSpanString);
    Utf8Parser.TryParse(timeSpanStringUtf8, out timeSpan, out _, standardFormat: 'c');
    Console.WriteLine(timeSpan);

    string json = $$"""
    {
      "Delay": "{{timeSpanString}}"
    }
    """;

    var options = JsonConvert.DeserializeObject<MyOptions>(json);
    Console.WriteLine(options?.Delay);

    try
    {
        options = System.Text.Json.JsonSerializer.Deserialize<MyOptions>(json);
        Console.WriteLine(options?.Delay);
    }
    catch (Exception e)
    {
        Console.WriteLine(e);
    }

    Console.WriteLine("----");
}
public class MyOptions
{
    public TimeSpan Delay { get; set; }
}
```

### Expected behavior

The above code should print each TimeSpan value in `timeSpanStrings` 4 times:

* TimeSpan.Parse
* Utf8Parser.TryParse
* The result of Newtonsoft.Json's `JsonConvert.DeserializeObject`
* The result from JsonSerializer.Deserialize

### Actual behavior

```
00:01:00
00:01:00
00:01:00
System.Text.Json.JsonException: The JSON value could not be converted to System.TimeSpan. Path: $.Delay | LineNumber: 1 | BytePositionInLine: 18.
 ---> System.FormatException: The JSON value is not in a supported TimeSpan format.
   at System.Text.Json.ThrowHelper.ThrowFormatException(DataType dataType)
   at System.Text.Json.Serialization.Converters.TimeSpanConverter.ReadCore(Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, Utf8JsonReader& reader, Exception ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at Program.<Main>$(String[] args) in C:\Users\eerhardt\source\repos\ConsoleApp115\ConsoleApp115\Program.cs:line 27
----
10.00:00:00
10.00:00:00
10.00:00:00
System.Text.Json.JsonException: The JSON value could not be converted to System.TimeSpan. Path: $.Delay | LineNumber: 1 | BytePositionInLine: 15.
 ---> System.FormatException: The JSON value is not in a supported TimeSpan format.
   at System.Text.Json.ThrowHelper.ThrowFormatException(DataType dataType)
   at System.Text.Json.Serialization.Converters.TimeSpanConverter.ReadCore(Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, Utf8JsonReader& reader, Exception ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at Program.<Main>$(String[] args) in C:\Users\eerhardt\source\repos\ConsoleApp115\ConsoleApp115\Program.cs:line 27
----
```

### Regression?

No

**Comment by eiriktsarpalis** [comment]
The built-in converter for `TimeSpan` is intentionally blocking out non-canonical representations. [This discussion](https://github.com/dotnet/runtime/pull/54186#discussion_r651698255) in the implementation PR should provide more context as to why we went for that route. TL;DR it was found that the `TimeSpan` parser was too permissive sometimes resulting in surprising results. Because the converter is capable of round tripping formats generated by `TimeSpan.ToString()` we decided to go ahead with this more restrictive version, and potentially relaxing it in the future provided there is demand.

**Comment by eerhardt** [comment]
> TL;DR it was found that the TimeSpan parser was too permissive sometimes resulting in surprising results. 

I can totally understand restricting to the `'c'` format to eliminate the surprising results, that makes sense. The part that I think should be fixed is adding further restrictions from the `'c'` format. Basically if `Utf8Parser.TryParse(timeSpanStringUtf8, out timeSpan, out _, standardFormat: 'c');` supports the format, so should JsonSerializer.

**Comment by danielmarbach** [comment]
We ran into a similar issue. Newtonsoft.Json supports both the 'c' format and the 'g' format (short version) but the built-in `TimeSpanConverter` restricts the [minimum length to be '8' ](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs#L11)even though `Utf8Parser.TryParse(timeSpanStringUtf8, out timeSpan, out _, standardFormat: 'c')` would accept with pleasure and parse also the shorter 7 digits version. 

Example

Newtonsoft.Json accepts `0:00:02` and `00:00:02` while the System.Text.Json converter only accepts `00:00:02`

**Comment by eiriktsarpalis** [comment]
> Basically if `Utf8Parser.TryParse(timeSpanStringUtf8, out timeSpan, out _, standardFormat: 'c');` supports the format, so should JsonSerializer.

The built-in converter does delegate to that routine when parsing:

https://github.com/dotnet/runtime/blob/9be628743487aac580a5e933f08d459dc50c80fb/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs#L60

Modulo rejecting leading whitespace the two should be identical.

**Comment by danielmarbach** [comment]
Except this check https://github.com/dotnet/runtime/blob/9be628743487aac580a5e933f08d459dc50c80fb/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs#L35 prevents things

**Comment by eiriktsarpalis** [comment]
Can you share examples of valid inputs that are caught by this particular check? This is there simply to avoid allocating an intermediate buffer (and multiplies the maximum expected size by a factor of 6 to account for potential escaped values)

**Comment by danielmarbach** [comment]
> Example
> 
> Newtonsoft.Json accepts `0:00:02` and `00:00:02` while the System.Text.Json converter only accepts `00:00:02`

See https://github.com/dotnet/runtime/issues/100538#issuecomment-2093104012

**Comment by danielmarbach** [comment]
The following program demonstrates things

```sharp
using System;
using JsonSerializer = System.Text.Json.JsonSerializer;
using Newtonsoft.Json;
					
public class Program
{
	public static void Main()
	{
		var timespan = TimeSpan.FromSeconds(2);
		var toSerialize = new ToSerialize() { SomeTimeSpan = timespan };
		
		Console.WriteLine(JsonConvert.DeserializeObject<ToSerialize>(@"{ ""SomeTimeSpan"": ""0:00:02"" }"));
		Console.WriteLine(JsonConvert.DeserializeObject<ToSerialize>(@"{ ""SomeTimeSpan"": ""00:00:02"" }"));
		
		Console.WriteLine(JsonSerializer.Deserialize<ToSerialize>(@"{ ""SomeTimeSpan"": ""0:00:02"" }"));
	}
	
	class ToSerialize {
	  public TimeSpan SomeTimeSpan { get; set; }	
	}
}
```

`Console.WriteLine(JsonSerializer.Deserialize<ToSerialize>(@"{ ""SomeTimeSpan"": ""0:00:02"" }"));`

fails with

```
Unhandled exception. System.Text.Json.JsonException: The JSON value could not be converted to System.TimeSpan. Path: $.SomeTimeSpan | LineNumber: 0 | BytePositionInLine: 27.
 ---> System.FormatException: The JSON value is not in a supported TimeSpan format.
   at System.Text.Json.ThrowHelper.ThrowFormatException(DataType dataType)
   at System.Text.Json.Serialization.Converters.TimeSpanConverter.ReadCore(Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, Utf8JsonReader& reader, Exception ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
```

**Comment by eiriktsarpalis** [comment]
> Example
>
> Newtonsoft.Json accepts 0:00:02 and 00:00:02 while the System.Text.Json converter only accepts 00:00:02

Right, but that input would be rejected on account of it not satisfying the `'c'` format, not because the input is too long.

**Comment by danielmarbach** [comment]
```csharp
Utf8Parser.TryParse(Encoding.UTF8.GetBytes("0:00:02"), out TimeSpan parsedTimeSpan, out _, standardFormat: 'c')
```
returns true

**Comment by eiriktsarpalis** [comment]
I see, so presumably it gets rejected because the input is too short? I think we should address that.

**Comment by eiriktsarpalis** [comment]
Happy to accept a PR that addresses this by the way, it's unlikely we would be able to prioritize this any time soon.

**Comment by danielmarbach** [comment]
> Happy to accept a PR that addresses this by the way, it's unlikely we would be able to prioritize this any time soon.

I see what I can do

**Comment by danielmarbach** [comment]
@eiriktsarpalis Here we go https://github.com/dotnet/runtime/pull/102091

// cc @eerhardt

**Comment by danielmarbach** [comment]
Will the fix be backported?

**Comment by eiriktsarpalis** [comment]
Unfortunately, no. Only high-impact bugs and security fixes get backported.
